src/lib/mongoBackend/mongoSubCache.cpp

```diff
@@ -0,0 +1,1204 @@
+)
+{
+  CachedSubscription* cSubP = mongoSubCache.head;
+
+  while (cSubP != NULL)
+  {
+    if (subMatch(cSubP, tenant, servicePath, entityId, entityType, attr))
+    {
+      subVecP->push_back(cSubP);
+      LM_T(LmtMongoSubCache, ("added subscription '%s': lastNotificationTime: %lu", cSubP->subscriptionId, cSubP->lastNotificationTime));
+    }
+
+    cSubP = cSubP->next;
+  }
+}
+
+
+
+/* ****************************************************************************
+*
+* cachedSubscriptionDestroy - 
+*/
+static void cachedSubscriptionDestroy(CachedSubscription* cSubP)
+{
+  if (cSubP->tenant != NULL)
+  {
+    free(cSubP->tenant);
+    cSubP->tenant = NULL;
+  }
+
+  if (cSubP->servicePath != NULL)
+  {
+    free(cSubP->servicePath);
+    cSubP->servicePath = NULL;
+  }
+
+  if (cSubP->subscriptionId != NULL)
+  {
+    free(cSubP->subscriptionId);
+    cSubP->subscriptionId = NULL;
+  }
+
+  if (cSubP->reference != NULL)
+  {
+    free(cSubP->reference);
+    cSubP->reference = NULL;
+  }
+
+  for (unsigned int ix = 0; ix < cSubP->entityIdInfos.size(); ++ix)
+  {
+    cSubP->entityIdInfos[ix]->release();
+    delete cSubP->entityIdInfos[ix];
+  }
+  cSubP->entityIdInfos.clear();
+
+  while (cSubP->attributes.size() > 0)
+  {
+    cSubP->attributes.erase(cSubP->attributes.begin());
+  }
+  cSubP->attributes.clear();
+
+  cSubP->notifyConditionVector.release();
+  cSubP->notifyConditionVector.vec.clear();
+
+  cSubP->next = NULL;
+}
+
+
+
+/* ****************************************************************************
+*
+* mongoSubCacheDestroy - 
+*/
+void mongoSubCacheDestroy(void)
+{
+  LM_T(LmtMongoSubCache, ("destroying subscription cache"));
+
+  CachedSubscription* cSubP  = mongoSubCache.head;
+
+  if (mongoSubCache.head == NULL)
+  {
+    return;
+  }
+
+  while (cSubP->next)
+  {
+    CachedSubscription* prev = cSubP;
+
+    cSubP = cSubP->next;
+    cachedSubscriptionDestroy(prev);
+    LM_T(LmtMongoSubCache,  ("removing CachedSubscription at %p", prev));
+    delete prev;
+  }
+
+  cachedSubscriptionDestroy(cSubP);
+  LM_T(LmtMongoSubCache,  ("removing CachedSubscription at %p", cSubP));
+  delete cSubP;
+
+  mongoSubCache.head  = NULL;
+  mongoSubCache.tail  = NULL;
+}
+
+
+
+/* ****************************************************************************
+*
+* tenantMatch - 
+*/
+static bool tenantMatch(const char* tenant1, const char* tenant2)
+{
+  //
+  // Removing complications with NULL, giving NULL tenants the value of an empty string;
+  //
+  tenant1 = (tenant1 == NULL)? "" : tenant1;
+  tenant2 = (tenant2 == NULL)? "" : tenant2;
+
+  if (strlen(tenant1) != strlen(tenant2))
+  {
+    return false;
+  }
+
+  if (strcmp(tenant1, tenant2) == 0)
+  {
+    return true;
+  }
+
+  return false;
+}
+
+
+
+/* ****************************************************************************
+*
+* mongoSubCacheItemLookup - 
+*
+* FIXME P7: lookups would be A LOT faster if the subCache used a hash-table instead of
```

Comment:
Of course.
However, it would be more visible if we created an issue about it.
We are very good looking at issues, but the FIXMEs ... not so good ...